cclunit-framework-source/doc/CCLUTMOCKING.md

```diff
@@ -0,0 +1,268 @@
+
+4. The mocking API calls cannot be used from reportwriter sections.  Alternatives would be dependent on the use-case, but take, for example, a dummyt used with a record structure and a call to cclutAddMockData within the detail section in order to add data based on a record structure.  Instead, a FOR loop construct could be leveraged outside the context of a reportwriter section to iterate over the record structure.
+
+5. Mocking record structures with a call to tdbexecute is unsupported under certain conditions, specifically if a call to free the record structure is made just prior to calling tdbexecute.  If the scenario is truly necessary for a test, the best alternative is to separate the freeing of the record structure and the call to tdbexecute in different subroutines and test the subroutines independently of each other.
+  
+## Example
+Below is an example of some of the APIs available in the CCL Unit Mocking framework along with some simple notes.
+
+Script-under-test:
+
+    drop program cclut_get_persons:dba go
+    create program cclut_get_persons:dba    
+    
+    declare newSize = i4 with protect, noconstant(0)
+    
+    select into "nl:"
+    from person p
+    plan p
+	order by p.person_id
+	detail
+		newSize = newSize + 1
+		stat = alterlist(reply->persons, newSize)
+		reply->persons[newSize].person_id = p.person_id
+		reply->persons[newSize].name_last = p.name_last
+		reply->persons[newSize].name_first = p.name_first
+		reply->persons[newSize].birth_dt_tm = p.birth_dt_tm
+	with nocounter
+    
+    end
+    go
+
+Test Code:
+
+    declare mockTable = vc with protect, noconstant("")
+    
+    ; Defining a mock person table.  The return value is the name of the mockTable.  
+    ; This can be useful to perform a select on the table after the script-under-test is complete
+    ; to verify (among other things) that an insert or a delete worked correctly.
+    set mockTable = cclutDefineMockTable("person", "person_id|name_last|name_first|birth_dt_tm",
+	    "f8|vc|vc|dq8")
+	
+	; Add a constraint that the person_id cannot be null
+	call cclutAddMockConstraint("person", "person_id", "not null")
+	
+	; Add a non-unique index to name_last
+	call cclutAddMockIndex("person", "name_last", 0)
+	
+	; Creates the mock table.  After this, it is available for DML statements.
+	call cclutCreateMockTable("person")
+	
+	; Create data for the table.
+	call cclutAddMockData("person", "1.0|Washington|George|01-JAN-1970 00:00") ;Will add George Washington 
+	call cclutAddMockData("person", "2.0|Adams|John|02-FEB-1971 11:11") ;Will add John Adams 
+	call cclutAddMockData("person", "3.0|Jefferson|\null|03-MAR-1972 22:22") ;Will add Jefferson (no first name) 
+	call cclutAddMockData("person", "4.0|Madison||04-APR-1973 10:33") ;Will add Madison (empty string for first name)
+	
+	record mock_reply (
```

Comment:
> I would be inclined to name this the_reply or cgp_reply rather than mock_reply. ...because this code isn't really mocking reply but rather emulating how some other script would execute cclut_get_persons if it had its own structure named reply.

Fair point.  I'll update that.

> Is there value in mentioning this subtlety in the cclutExecuteScriptWithMocks documentation ? ... that one can effect with replace("A", b) by adding "b" as a mock for "a".

Yes, I think that does seem worthwhile to call out explicitly.  I'll add that.